pkg/api/vmtemplate/store.go

```diff
@@ -0,0 +1,138 @@
+	}
+
+	templateNs, templateName := ref.Parse(templateID)
+	if ns != templateNs {
+		return types.APIObject{}, apierror.NewAPIError(validation.InvalidBodyContent, "Template version and template should belong to same namespace")
+	}
+
+	if err := s.setOwnerRefAndLabel(request, data); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.InvalidBodyContent, err.Error())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	sets := labels.Set{
+		templateLabel: templateName,
+	}
+	existedVersions, err := s.templateVersions.List(ns, metav1.ListOptions{LabelSelector: sets.AsSelector().String()})
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, err.Error())
+	}
+
+	currentVersion := len(existedVersions.Items) + 1
+	data.Data().SetNested(currentVersion, "status", "versionNumber")
+	data.Data().SetNested(templateName+"-", "metadata", "generateName")
+
+	return s.Store.Create(request, request.Schema, data)
+}
+
+func (s *templateVersionStore) Update(request *types.APIRequest, schema *types.APISchema, data types.APIObject, id string) (types.APIObject, error) {
+	return types.APIObject{}, apierror.NewAPIError(validation.ActionNotAvailable, "Update templateVersionCache is not supported")
+}
+
+func (s *templateVersionStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	if err := s.canDeleteTemplateVersion(id); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, err.Error())
+	}
+
+	return s.Store.Delete(request, request.Schema, id)
+}
+
+func (s *templateVersionStore) canDeleteTemplateVersion(id string) error {
+	ns, name := ref.Parse(id)
+	vr, err := s.templateVersionCache.Get(ns, name)
+	if err != nil {
+		return err
+	}
+
+	vtNS, vtname := ref.Parse(vr.Spec.TemplateID)
+	vt, err := s.templateCache.Get(vtNS, vtname)
+	if err != nil {
+		return err
+	}
+
+	if vt.Spec.DefaultVersionID == id {
+		return errors.New("Cannot delete the default templateVersion")
+	}
+
+	return nil
+}
+
+func (s *templateVersionStore) setOwnerRefAndLabel(request *types.APIRequest, data types.APIObject) error {
+	templateID := data.Data().String("spec", "templateId")
+	if templateID == "" {
+		return errors.New("TemplateID shouldn't be empty")
+	}
+	ns, name := ref.Parse(templateID)
+	template, err := s.templateCache.Get(ns, name)
+	if err != nil {
+		return err
+	}
+
+	var ownerReferencesSlice []map[string]interface{}
+	ownerReference := map[string]interface{}{
+		"kind":               "vmtemplate",
+		"apiVersion":         template.APIVersion,
+		"name":               template.Name,
+		"uid":                template.UID,
+		"blockOwnerDeletion": true,
+		"controller":         true,
+	}
+	ownerReferencesSlice = append(ownerReferencesSlice, ownerReference)
+
+	data.Data().SetNested(ownerReferencesSlice, "metadata", "ownerReferences")
+	data.Data().SetNested(name, "metadata", "labels", templateLabel)
+	return nil
```

Comment:
what if user created the VM template and version through the `kubectl`, probably should move the logic to the controller?

Reply:
moved to controller